File: BitManipulation/Python/_0191_Numberof1Bits.py
class Solution:
    # Iteration
    # Time Complexity: BigO(N)
    # Space Complexity: BigO(1)
    def hammingWeight(self, n: int) -> int:
        count = 0
        while n:
            n &= n - 1
            count += 1
        return count

    # Iteration
    # Time Complexity: BigO(N)
    # Space Complexity: BigO(1)
    def hammingWeightV2(self, n: int) -> int:
        count = 0
        while n:
            count += n & 1
            n >>= 1
        return count

    # Iteration
    # Time Complexity: BigO(1)
    # Space Complexity: BigO(1)
    def hammingWeightV3(self, n: int) -> int:
        return bin(n).count('1')

    # Iteration
    # Time Complexity: BigO(1)
    # Space Complexity: BigO(1)
    def hammingWeightV4(self, n: int) -> int:
        return n.bit_count()


ins = Solution()

print(ins.hammingWeightV2(11))
print(ins.hammingWeightV2(128))
print(ins.hammingWeightV2(4294967293))

File: BitManipulation/TypeScript/_0191_Numberof1Bits.ts
/**
 * Iteration
 * Time Complexity: BigO(N)
 * Space Complexity: BigO(1)
 *
 * Instead of using reduce(), filter() can be used
 */
function hammingWeight(n: number): number {
  return n
    .toString(2)
    .split("")
    .reduce((acc, cur) => {
      if (cur === "1") return (acc += 1);
      else return (acc += 0);
    }, 0);
}

console.log(hammingWeight(11));
console.log(hammingWeight(128));
console.log(hammingWeight(4294967293));
